### PR TITLE
[FLINK-31780][component=Runtime/Coordination] Allow users to disable 'Ensemble tracking' feature for ZooKeeper Curator framework

### PR DESCRIPTION
**Pull request by JTaky**
## What is the purpose of the change

Given PR adds a new configuration to disable EnsembleTracking feature for Apache Curator. It would make possible to actually configure HA for Zookeeper, which is living behind VIP.

Currently HA with ZooKeeper can even lead to the JobManager failure. The scenario of the failure is next:
1. Flink connects to ZooKeeper via configured URL.
2. Ensemble tracking gets a new URL of ensemble, which is not obligatory accessible for Flink, because Zookeeper is under VIP.
3. In case of reconnect Flink fails to Zookeeper, moreover due to "UnresolvedHostException" Flink's jobManager is killed.

## Brief change log

- Adds a new configuration into HighAvailabilityOptions
- Use newly added ZOOKEEPER_ENSEMBLE_TRACKING flag to control Curator ensemble tracking
- Default value is enabling ensemble tracking for backward compatibility, see [Curator source code](https://github.com/apache/curator/blob/master/curator-framework/src/main/java/org/apache/curator/framework/CuratorFrameworkFactory.java#L72)

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage. 
The code path is executed in several test scenarios, e.g. ZooKeeperLeaderElectionITCase#testJobExecutionOnClusterWithLeaderChange.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: yes
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? docs / JavaDocs

### CODE REVIEW

File: flink-core/src/main/java/org/apache/flink/configuration/HighAvailabilityOptions.java [hunk 1]
```diff
@@ -199,6 +199,23 @@ public class HighAvailabilityOptions {
...
+    public static final ConfigOption<Boolean> ZOOKEEPER_ENSEMBLE_TRACKING =
+            key("high-availability.zookeeper.client.ensemble-tracker")
+                    .booleanType()
+                    .defaultValue(true)
+                    .withDescription(
+                            Description.builder()
+                                    .text(
+                                            "Defines whether Curator should enable ensemble tracker. This can be useful in certain scenarios "
+                                                    + "in which CuratorFramework is accessing to ZK clusters via load balancer or Virtual IPs. "
+                                                    + "Default Curator EnsembleTracking logic watches CuratorEventType.GET_CONFIG events and"
```
Comment (XComp) [inline flink-core/src/main/java/org/apache/flink/configuration/HighAvailabilityOptions.java:212]:
```suggestion
                                                    + "Default Curator EnsembleTracking logic watches CuratorEventType.GET_CONFIG events and "
```

File: flink-core/src/main/java/org/apache/flink/configuration/HighAvailabilityOptions.java [hunk 2]
```diff
@@ -199,6 +199,23 @@ public class HighAvailabilityOptions {
...
+                                                    + "changes zookeeper connection string, it is not desired behaviour when ZooKeeper is running under the Virtual IPs."
```
Comment (XComp) [inline flink-core/src/main/java/org/apache/flink/configuration/HighAvailabilityOptions.java:213]:
```suggestion
                                                    + "changes ZooKeeper connection string. It is not desired behaviour when ZooKeeper is running under the Virtual IPs. "
```

File: flink-core/src/main/java/org/apache/flink/configuration/HighAvailabilityOptions.java [hunk 3]
```diff
@@ -199,6 +199,23 @@ public class HighAvailabilityOptions {
...
+                                                    + "Under certain configurations EnsembleTracking can lead to setting of Zookeeper connection string "
```
Comment (XComp) [inline flink-core/src/main/java/org/apache/flink/configuration/HighAvailabilityOptions.java:214]:
```suggestion
                                                    + "Under certain configurations EnsembleTracking can lead to setting of ZooKeeper connection string "
```

File: flink-core/src/main/java/org/apache/flink/configuration/HighAvailabilityOptions.java [hunk 4]
```diff
@@ -199,6 +199,23 @@ public class HighAvailabilityOptions {
...
+                                                    + "with unresolvable hostnames.",
+                                            TextElement.code("true"))
```
Comment (XComp) [inline flink-core/src/main/java/org/apache/flink/configuration/HighAvailabilityOptions.java:216]:
```suggestion
                                                    + "with unresolvable hostnames.")
```
Looks like a copy&paste error.

**Review by XComp — CHANGES_REQUESTED** [review]
Thanks for the PR, @JTaky . I have a few minor comments. PTAL

**Comment by XComp** [comment]
CI is still failing.

**Comment by JTaky** [comment]
@flinkbot run azure

**Review by XComp — APPROVED** [review]
Thanks for the change, @JTaky . I squashed the commits together and rebased the branch (I wasn't sure how Github's squash-and-merge works with the merge commits. I need to try that out for the future.)

**Comment by XComp** [comment]
@flinkbot run azure

**Comment by JTaky** [comment]
Thanks a lot for the review! 
Sorry but I think I cannot merge this PR.

Screenshot 2023-05-12 at 18 33 32

**Pull request merged**